// @wdio/cucumber-framework/build/hookRunner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _config = require("@wdio/config");

var _cucumberEventListener = _interopRequireDefault(require("./cucumberEventListener"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class HookRunner {
  constructor(eventBroadcaster, config) {
    _defineProperty(this, "gherkinDocEvents", []);

    this.config = config;
    this.eventListener = new _cucumberEventListener.default(eventBroadcaster).on('before-feature', this.handleBeforeFeature.bind(this)).on('before-scenario', this.handleBeforeScenario.bind(this)).on('before-step', this.handleBeforeStep.bind(this)).on('after-step', this.handleAfterStep.bind(this)).on('after-scenario', this.handleAfterScenario.bind(this)).on('after-feature', this.handleAfterFeature.bind(this));
  }

  handleBeforeFeature(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.beforeFeature, args);
  }

  handleBeforeScenario(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.beforeScenario, args);
  }

  handleBeforeStep(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.beforeStep, args);
  }

  handleAfterStep(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.afterStep, args);
  }

  handleAfterScenario(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.afterScenario, args);
  }

  handleAfterFeature(...args) {
    return (0, _config.executeHooksWithArgs)(this.config.afterFeature, args);
  }

}

exports.default = HookRunner;
